csrc/preseg_passes/factor_amax_reduction.cpp

```diff
@@ -0,0 +1,508 @@
+  valid_states.emplace(Broadcast, std::vector<State>({MaxReduction}));
+  valid_states.emplace(Cast, std::vector<State>({Broadcast, MaxReduction}));
+  valid_states.emplace(MaxReduction, std::vector<State>({Abs}));
+
+  // Initial state
+  TensorView* max_reduction_tv = nullptr;
+  TensorView* current_tv = tv;
+  State current_state = Start;
+  State next_state = Start;
+
+  if (isDebugDumpEnabled(DebugDumpOption::PreSegmenterLogging)) {
+    debug() << "==== Running detectAmaxPattern " << tv->toString()
+            << " ====" << std::endl;
+  }
+
+  while (current_tv != nullptr) {
+    // Get potential next state based on TensorView definition
+    if (findUnaryDefinition(current_tv, UnaryOpType::Cast)) {
+      next_state = Cast;
+    } else if (findUnaryDefinition(current_tv, UnaryOpType::Abs)) {
+      next_state = Abs;
+      break;
+    } else if (findReductionDefinition(current_tv, BinaryOpType::Max)) {
+      next_state = MaxReduction;
+    } else if (current_tv->definition()->isA<BroadcastOp>()) {
+      next_state = Broadcast;
+    } else {
+      next_state = Invalid;
+    }
+
+    if (isDebugDumpEnabled(DebugDumpOption::PreSegmenterLogging)) {
+      debug() << "From\t" << current_state << "\tTo\t" << next_state
+              << std::endl;
+    }
+
+    // Validate next state given current state
+    bool is_valid = (valid_states.count(current_state) == 0) ||
+        std::any_of(valid_states.at(current_state).begin(),
+                    valid_states.at(current_state).end(),
+                    [&](State s) { return s == next_state; });
+    if (!is_valid) {
+      next_state = Invalid;
+    }
+
+    // Transition to next state
+    switch (next_state) {
+      case Start:
+      case Invalid: {
+        // For these failure states, we clear max_reduction_tv.
+        max_reduction_tv = nullptr;
+        [[fallthrough]];
+      }
+      case Abs: {
+        // For the pass state, we return max_reduction_tv.
+        current_tv = nullptr;
+        break;
+      }
+      case MaxReduction: {
+        // Set max_reduction_tv
+        max_reduction_tv = current_tv;
+        [[fallthrough]];
+      }
+      case Broadcast:
+      case Cast: {
+        // Move to input TensorView of definition
+        current_tv = current_tv->definition()->input(0)->as<TensorView>();
+        break;
+      }
+    }
+    current_state = next_state;
+  }
+
+  if (isDebugDumpEnabled(DebugDumpOption::PreSegmenterLogging)) {
+    debug() << "==== Finish detectAmaxPattern ====" << std::endl;
+  }
+
+  NVF_ERROR(
+      max_reduction_tv == nullptr ||
+      findReductionDefinition(max_reduction_tv, BinaryOpType::Max));
+  return max_reduction_tv;
+}
+
+// Select a partial reduction TensorView to factor the amax reduction.
+// Given the lack of information in the presegmentation pass, we select the
+// partial reduction with shortest dependency chain with amax reduction.
+TensorView* findUpstreamReduction(Fusion* fusion, TensorView* amax_reduction) {
```

Comment:
I think we need to pay more attention to the topology here. Once we found the closest reduction producer, I think we need to ensure that no TV will introduce more concrete IterDomain into the chain on the way to the full reduction.

```
TV1 = reducction(TV0)
TVn = follow_up_operations(TV0, TV1, ..., possibly other TVs as well)
amax(TVn);
```

Does it make sense to require that TVn's logical domain being a subset of the logical domain of TV0. So we know that TVn would be a valid reference tensor for the reduction?

This doesn't feel optimal, I'm wondering if it would help us get rid of the regresssion.

Reply:
I think if there is a broadcast post-reduction, we need to check for a persistent pattern. Otherwise, the reduction scheduler will reject this pattern.